services.py

```diff
@@ -110,6 +131,84 @@ def update_or_create_criterion(data, criterion=None):
     return criterion.save()
 
 
+def update_or_create_score(criterion_id, state_code, meets_criterion):
+    '''
+    Takes a criterion id, a state code, and whether the state meets that criterion.
+    If the current score for this state/criterion does not match what is provided,
+    or no score exists, a new score is created with this information.
+    '''
+    score = Score.query.filter_by(criterion_id=criterion_id, state=state_code) \
+        .order_by(Score.created_at.desc()).first()
+
+    if score is None or score.meets_criterion != meets_criterion:
+        score = Score(
+            criterion_id=criterion_id,
+            state=state_code,
+            meets_criterion=meets_criterion,
+        ).save()
+
+    return score
+
+
+def update_or_create_state_category_grade(category_id, state_code, grade):
+    '''
+    Takes a category id, a state code, and the grade for that state and category.
+    If the current grade for this state/category does not match what is provided, or if no grade
+    exists, a new grade is created with this information.
+    '''
+    state_category_grade = StateCategoryGrade.query.filter_by(
+        state_code=state_code,
+        category_id=category_id,
+    ).order_by(StateCategoryGrade.created_at.desc()).first()
+
+    if state_category_grade is None or state_category_grade.grade != grade:
+        state_category_grade = StateCategoryGrade(
+            state_code=state_code,
+            category_id=category_id,
+            grade=grade,
+        ).save()
+
+    return state_category_grade
+
+
+def update_state(data, state):
```

Comment:
This is a bunch of logic for updating state data from a JSON payload. It's heavily tested in the file below this one :)